### PR TITLE
Fix race conditions in task callback invocations

### PR DESCRIPTION
**Pull request by houqp**
This race condition resulted in task success and failure callbacks being
called more than once. Here is the order of events that could lead to
this issue:

* task started running within process 2
* (process 1) local_task_job checked for task return code, returns None
* (process 2) task exited with failure state, task state updated as failed in DB
* (process 2) task failure callback invoked through taskinstance.handle_failure method
* (process 1) local_task_job heartbeat noticed task state set to
  failure, mistoken it as state bing updated externally, also invoked task
  failure callback

To avoid this race condition, we need to make sure task callbacks are
only invoked within a single process.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by houqp** [comment]
there is one test that i forgot to update, will update it in couple hours.

File: airflow/jobs/local_task_job.py [hunk 1]
```diff
@@ -162,11 +162,7 @@ def heartbeat_callback(self, session=None):
                 "Terminating instance.",
                 ti.state
             )
-            if ti.state == State.FAILED and ti.task.on_failure_callback:
-                context = ti.get_template_context()
-                ti.task.on_failure_callback(context)
-            if ti.state == State.SUCCESS and ti.task.on_success_callback:
-                context = ti.get_template_context()
-                ti.task.on_success_callback(context)
```
Comment (turbaszek) [inline airflow/jobs/local_task_job.py:177]: In case of success, do I correctly understand that now `on_success_callback` will be executed only in `_run_raw_task` via `_run_success_callback`?

Comment (houqp) [reply airflow/jobs/local_task_job.py:177]: that's correct, callbacks should only be executed within _run_raw_task to avoid race condiitions.

**Review by turbaszek — APPROVED** [review]

**Review by kaxil — APPROVED** [review]
LGTM.

**Comment by kaxil** [comment]
cc @mik-laj @ashb You might want to take a look at it

**Review by ashb — COMMENTED** [review]
This sounds broadly good -- except there's one case not handled:

if the raw task process _dies hard_ (segfault, OOM killed say) then the failure callback wouldn't be executed.

**Comment by houqp** [comment]
@ashb this is the tradeoff we will have to make here, segfault, OOM kill could happen to local_scheduler_job as well, although the chance is much lower. Either way, we need to make sure callbacks are only invoked from within one process, it's either the scheduler job, or raw_task, but not from both. This is happening in production for us at a high frequency right now, duplicated failure callbacks are invoked multiple times a day, while I don't recall running into hard die scenarios, so I think it's a reasonable tradeoff.

On top of this, the refactoring here only changes behavior of callback invocation triggered by external state change. It's very unlikely that external state are updated right before task got into OOM or segfault. With or without this change, Airflow still invokes failure callback from within raw_task when state are not changed externally, so the problem you mentioned already exists in today's code base.

From a design's point of view, it's better to invoke success and failure callbacks from the task monitor process, e.g. the local scheduler job, but it would require a much bigger refactoring. If that's what the community prefers, I can give that a stab. The run task command needs to be aware of whether it's been invoked with or without an external task monitor, and change the callback invocation logic based on that.

**Comment by houqp** [comment]
changing title back to WIP since I am going to do a big refactor to move all success/callback invocations into caller of _run_raw_task.

File: airflow/jobs/local_task_job.py [hunk 2]
```diff
@@ -170,11 +174,6 @@ def heartbeat_callback(self, session=None):
...
             self.task_runner.terminate()
+            ti.run_finished_callback()
```
Comment (houqp) [inline airflow/jobs/local_task_job.py]: changed the semantic to kill the task first before running callback, it would be odd to execute success callback while task is still running.

File: airflow/models/dag.py [hunk 3]
```diff
@@ -561,7 +561,7 @@ def next_dagrun_after_date(self, date_last_automated_dagrun: Optional[pendulum.D
         next_run_date = None
         if not date_last_automated_dagrun:
             # First run
-            task_start_dates = [t.start_date for t in self.tasks]
+            task_start_dates = [t.start_date for t in self.tasks if t.start_date]
```
Comment (houqp) [inline airflow/models/dag.py:587]: drive by bug fix for lint error, t.start_date could be None

File: airflow/models/taskinstance.py [hunk 4]
```diff
@@ -1345,11 +1354,12 @@ def handle_failure(self, error, test_mode=None, context=None, force_fail=False,
         # Log failure duration
         session.add(TaskFail(task, self.execution_date, self.start_date, self.end_date))
 
-        if context is not None:
-            context['exception'] = error
```
Comment (houqp) [inline airflow/models/taskinstance.py:1409]: this is the biggest semantic change this PR introduces, failure callback won't have access to `context['exception']` anymore because it's not called from a separate task manager process. It is unavoidable because if we don't execute callback from external monitor process, then we won't be able to catch task failures caused by out of memory or segfault.

I think this is a reasonable tradeoff because some of the success/failure callbacks were already invoked from external manager process when states are updated externally, therefore, the contract for accessing `context['exception']` is already partially broken today.

File: tests/jobs/test_local_task_job.py [hunk 5]
```diff
@@ -438,5 +521,5 @@ def test_number_of_queries_single_loop(self, mock_get_task_runner, return_codes)
         mock_get_task_runner.return_value.return_code.side_effects = return_codes
 
         job = LocalTaskJob(task_instance=ti, executor=MockExecutor())
-        with assert_queries_count(12):
+        with assert_queries_count(13):
```
Comment (houqp) [inline tests/jobs/test_local_task_job.py:523]: added a new DB query in local_task_job to query task state after task finished execution so we can decide to which callback to invoke

**Comment by houqp** [comment]
@ashb I have updated the PR to invoke callbacks from external task monitor code, please feel free to take a look when you have time.

@madison-ookla FYI in case you want to test this fix in your setup.

**Comment by houqp** [comment]
@kaxil added docs to updating.md as we discussed during the call today.

File: airflow/models/taskinstance.py [hunk 6]
```diff
@@ -1446,18 +1454,21 @@ def handle_failure(self, error, test_mode=None, context=None, force_fail=False,
...
             session.merge(self)
         session.commit()
 
+    @provide_session
+    def handle_failure_with_callback(
+        self,
+        error,
+        test_mode=None,
+        force_fail: bool = False,
+        session=None,
```
Comment (turbaszek) [inline airflow/models/taskinstance.py]: Should we add type hints?

Comment (houqp) [reply airflow/models/taskinstance.py]: @turbaszek added type hints for error and test_mode, but leaving out session since I am not sure what's the best way to handle because it's not supposed to be `Optional[Session]` due to use of provide_session decorator. But on the other hand we need to set the default value to `None` so the provide_session operator knows when to inject the session.

**Review by turbaszek — APPROVED** [review]

File: UPDATING.md [hunk 7]
```diff
@@ -94,6 +94,15 @@ to 'excluded/'.
 As [recommended](https://flask.palletsprojects.com/en/1.1.x/config/#SESSION_COOKIE_SAMESITE) by Flask, the
 `[webserver] cookie_samesite` has bee changed to `Lax` from `None`.
 
+### Breaking change to task failure callback
+
+To fix race conditions and make task exit state update more reliable, we moved
+task success/failure callback invocation into external monitoring processess.
+As a result, `context['exception']` is no longer accessible within task failure
```
Comment (ashb) [inline UPDATING.md]: Waaait, the `exception` is now _never_ available? That's far from ideal.

File: airflow/models/taskinstance.py (hunk 4 above)
Comment (ashb) [reply airflow/models/taskinstance.py:1409]: Sorry, this makes it a -1 from me. This information is vital when it's available and used by a number of people to send exception messages in Slack hooks. Without this it's much less useful

We need to keep this when it is available, and either document that it's not available in all cases, or add a "synthetic" exception when it's missing (`AirflowError("state changed to failed")` for example)

**Review by ashb — DISMISSED** [review]
Sorry, not happy about removing the exception from the context.

File: airflow/models/taskinstance.py [hunk 8]
```diff
@@ -1308,6 +1296,29 @@ def _run_execute_callback(self, context, task):
             self.log.error("Failed when executing execute callback")
             self.log.exception(exc)
 
+    def run_finished_callback(self) -> None:
```
Comment (ashb) [inline airflow/models/taskinstance.py]:
```suggestion
    def _run_finished_callback(self) -> None:
```

Since it should be "private"

**Comment by houqp** [comment]
> Sorry, not happy about removing the exception from the context.

We need to make a call here then, the trade off boils down to:

* handle state update from external process to cover hard crashes like OOM and segfaults, but provide a generic state changed to failed exception to failure_callback
* handle state update within run_raw_task, and live with the fact that hard crash will result in miss of task instance state clean up and callbacks

In theory, we can pass exception info from run_raw_task into the external monitoring process, but this will require a much bigger refactor if we want to use shared memory. Alternatively, we can write out exception info from run_raw_task into a local temp file to be picked up by external monitoring process.

What do you think? I am leaning towards the temp file solution, which seems to cover all requirements we have so far.

**Comment by ashb** [comment]
Yeah, for now a temp file that is created by the monitor process and passed down to the raw runner (which then probably uses pickle to store the exception in) sounds like a good approach for now.

Sorry to make "last minute" demands @houqp, but I think this is important to keep.

**Comment by turbaszek** [comment]
Not sure, but won't we be able to use `traceback` module? That's at least what I used for custom failure callback when I wanted more info.
```
exception = sys.exc_info()
traceback = "".join(traceback.format_exception(*exception))
```

**Comment by houqp** [comment]
@ashb no worry, I wasn't aware `context['exception']` is being widely used since this feature is currently half broken. I will refactor the code to exchange crash exception info through temp local file. But I would prefer to not use pickle and just dump the text since I think p

@turbaszek the reason we can't use traceback module here is because failure call site is now in an external monitoring process, which is an entirely different process compared to the one that runs the actual task. As a result, caller won't have access to the exception object in its memory.

@kaxil @ashb now that we won't be introduce a breaking change, I think we don't need to drag beta1 release along for this fix, we can include it in beta2 if needed.

**Comment by ashb** [comment]
@houqp the `ExceptionWithTraceback` we use in CeleryExecutor may be useful here - that is pickleable too.

File: airflow/cli/cli_parser.py [hunk 9]
```diff
@@ -396,6 +396,7 @@ def positive_int(value):
     ("--ship-dag",), help="Pickles (serializes) the DAG and ships it to the worker", action="store_true"
 )
 ARG_PICKLE = Arg(("-p", "--pickle"), help="Serialized pickle object of the entire dag (used internally)")
+ARG_EXCEPTION_FILE = Arg(("--error-file",), help="File to store task failure error")
```
Comment (ashb) [inline airflow/cli/cli_parser.py]: Error vs exception :)

File: airflow/models/baseoperator.py [hunk 10]
```diff
@@ -1102,7 +1102,8 @@ def run(
         end_date = end_date or self.end_date or timezone.utcnow()
 
         for execution_date in self.dag.date_range(start_date, end_date=end_date):
-            TaskInstance(self, execution_date).run(
+            ti = TaskInstance(self, execution_date)
+            ti.run(
```
Comment (ashb) [inline airflow/models/baseoperator.py]: Is this needed?

File: airflow/models/taskinstance.py [hunk 11]
```diff
@@ -1392,14 +1425,26 @@ def _handle_reschedule(self, actual_start_date, reschedule_exception, test_mode=
...
         """Handle Failure for the TaskInstance"""
         if test_mode is None:
             test_mode = self.test_mode
-        if context is None:
-            context = self.get_template_context()
 
-        self.log.exception(error)
+        if error:
+            self.log.exception(error)
+            # external monitoring process provides pickel file so _run_raw_task
```
Comment (ashb) [inline airflow/models/taskinstance.py]:
```suggestion
            # external monitoring process provides pickle file so _run_raw_task
```

File: airflow/models/taskinstance.py (hunk 4 above)
Comment (ashb) [reply airflow/models/taskinstance.py:1409]: We can remove this now can't we?

File: airflow/cli/cli_parser.py (hunk 9 above)
Comment (houqp) [reply airflow/cli/cli_parser.py]: i initially stated with exception, then I noticed it is named as error as in `taskinstance.handle_failure` and it's type could be a str or exception :( so i used error everywhere to be consistent with the type. I think error would be more accurate where, but let me know if you still prefer exception.

File: airflow/models/taskinstance.py [hunk 12]
```diff
@@ -106,6 +108,29 @@ def set_current_context(context: Context):
...
+    except Exception:  # pylint: disable=broad-except
+        return "Failed to load task run error"
+
+
+def set_error_file(error_file: str, error: Union[str, Exception]) -> None:
+    """Write error into error file by path"""
+    with open(error_file, "wb") as fd:
+        try:
+            pickle.dump(error, fd)
+        except Exception:  # pylint: disable=broad-except
```
Comment (houqp) [inline airflow/models/taskinstance.py:127]: @ashb minor change in behavior here. if user raises an exception from locally scoped class definition during task execution, we won't be able to pickle it, so we will have to serialize it as string instead. however, this should not be a breaking change since `context['exception']` has always been carrying a union type of `str` and `Exception`, so user code should not be expecting it to always be an `Exception` anyway.

File: airflow/models/taskinstance.py (hunk 4 above)
Comment (houqp) [reply airflow/models/taskinstance.py:1409]: yes, `context['exception']` is not set in `_run_finished_callback`.

File: airflow/cli/cli_parser.py (hunk 9 above)
Comment (ashb) [reply airflow/cli/cli_parser.py]: I just mean the variable name is EXCEPTION, but the flag is "error"

Comment (kaxil) [reply airflow/cli/cli_parser.py]: @houqp Can you address this please. Just a variable name change or description change :)

File: airflow/executors/debug_executor.py [hunk 13]
```diff
@@ -76,10 +77,14 @@ def _run_task(self, ti: TaskInstance) -> bool:
         try:
             params = self.tasks_params.pop(ti.key, {})
             ti._run_raw_task(job_id=ti.job_id, **params)  # pylint: disable=protected-access
+            ti.set_state(State.SUCCESS)
```
Comment (kaxil) [inline airflow/executors/debug_executor.py]: Why is this needed? How was it working before this PR?

File: airflow/cli/cli_parser.py (hunk 9 above)
Comment (houqp) [reply airflow/cli/cli_parser.py]: oh my bad, not sure why github is not notifying this comment to me earlier :(

File: airflow/executors/debug_executor.py (hunk 13 above)
Comment (houqp) [reply airflow/executors/debug_executor.py]: good catch, run_raw_task does update state to SUCCESS internally, it was causing issues in previous implementation of the PR, but we have changed the design a lot since then, so maybe it's not needed anymore. let me remove it and rerun the tests to see if anything breaks.

Comment (houqp) [reply airflow/executors/debug_executor.py]: looks like we don't need it anymore, tests passed.

**Review by turbaszek — APPROVED** [review]

**Comment by potiuk** [comment]
I guess this one MUST be merged before RC1 :)

File: airflow/models/taskinstance.py [hunk 14]
```diff
@@ -105,6 +107,29 @@ def set_current_context(context: Context):
             )
 
 
+def load_error_file(fd: IO[bytes]) -> Optional[Union[str, Exception]]:
+    """Load and return error from error file"""
+    fd.seek(0, 0)
```
Comment (ashb) [inline airflow/models/taskinstance.py]:
```suggestion
    fd.seek(0, os.SEEK_SET)
```
(Not important if we don't do this)

File: airflow/jobs/local_task_job.py [hunk 15]
```diff
@@ -98,7 +98,10 @@ def signal_handler(signum, frame):
 
             heartbeat_time_limit = conf.getint('scheduler', 'scheduler_zombie_task_threshold')
 
-            while True:
+            # task callback invocation happens either here or in
+            # self.heartbeat() instead of taskinstance._run_raw_task to
+            # avoid race conditions
+            while not self.terminating:
```
Comment (ashb) [inline airflow/jobs/local_task_job.py:108]: What's this change for? How come we didn't do it before?

**Review by ashb — COMMENTED** [review]
One question, one comment, but otherwise LGTM.

**Comment by kaxil** [comment]
All MySQL jobs timed out 🤔

**Comment by kaxil** [comment]
Re-running it just to be sure

**Comment by potiuk** [comment]
> Re-running it just to be sure

Sounds like no-accident :(

**Comment by kaxil** [comment]
@houqp Can you double-check please what's causing MySQL to hang

**Comment by ashb** [comment]
I have just rebased this on to latest master. 🤞🏻

**Comment by kaxil** [comment]
hmn no still same issue with MySQL

**Comment by houqp** [comment]
let me try to reproduce this on my end.

**Comment by houqp** [comment]
OK, I see what's going on, it's caused by the tempfile patching in `test_mysql_to_hive` test. I will push a fix for that test shortly.

File: airflow/models/taskinstance.py (hunk 14 above)
Comment (houqp) [reply airflow/models/taskinstance.py]: this is important, no magic number allowed :)

File: airflow/jobs/local_task_job.py (hunk 15 above)
Comment (houqp) [reply airflow/jobs/local_task_job.py:108]: This is needed because of newly added `self.handle_task_exit(return_code)`. If we restart the loop after heartbeat sets `self.terminating` to True, then handle_task_exit will be invoked which will result in duplicated callback invocations. I added more comment in the code to document this.

**Comment by kaxil** [comment]
Build docs step is failing -- rebased to latest master -- hopefully that should be fixed too

**Comment by houqp** [comment]
docs step is passing now, but one of the mysql integration test failed due to infra error, i am going to restart the pipeline.

**Comment by kaxil** [comment]
Some tests were failing with 137 error, just restarted them.

**Comment by houqp** [comment]
looks like another 2 tests failed with 137 error, let me try to restart again

**Comment by houqp** [comment]
@ashb all tests are passing now, we have been running this patch in production for couple weeks. do you want to do another round of review?

**Comment by ashb** [comment]
I'll take a look. We've been running with this at Astronomer too and haven't had any problems reported either

**Review by ashb — APPROVED** [review]

**Comment by kaxil** [comment]
@houqp Can you please rebase on Master one last time :) -- Thanks

**Comment by houqp** [comment]
@kaxil rebased :)
